crates/ra_parser/src/grammar/items.rs

```diff
@@ -161,14 +151,19 @@ pub(super) fn maybe_item(p: &mut Parser, flavor: ItemFlavor) -> MaybeItem {
         // unsafe default impl Foo {}
         IMPL_KW => {
             traits::impl_block(p);
-            IMPL_BLOCK
+            m.complete(p, IMPL_BLOCK);
+            None
         }
         _ => {
-            return if has_mods { MaybeItem::Modifiers } else { MaybeItem::None };
+            if has_mods {
+                p.error("expected fn, trait or impl");
+                m.complete(p, ERROR);
+                None
+            } else {
+                Some(m)
```

Comment:
Instead of lislint `None` in every branch, I think we can use `return Some` here, and then just a single `None` at the end of the function